### ISSUE TITLE
r3.0 release

### ISSUE THREAD

**Issue by meator** [enkore/j4-dmenu-desktop#161]
This issue tracks everything blocking the `r3.0` release.

# Fixes, changes
- [x] Fix bug causing segfaults in `NotifyInotify`
      Couldn't reproduce. I hope that it's working properly. It could be related (and could have been fixed by) 54bb50947200c03d4cd33e61a0e73270ff566f75
- [x] Update `CHANGELOG`
- [x] Remove `complgen` generation of Bash completion
      The idea was really nice on paper, but writing the completion script manually is the better solution. The current state of shell completions in modern shells is subpar IMO, but it is how it is.
- [x] Verify that [README.md](https://github.com/enkore/j4-dmenu-desktop/blob/develop/README.md), [CONTRIBUTING.md](https://github.com/enkore/j4-dmenu-desktop/blob/develop/CONTRIBUTING.md), [BUILDING.md](https://github.com/enkore/j4-dmenu-desktop/blob/develop/BUILDING.md) and [PACKAGING.md](https://github.com/enkore/j4-dmenu-desktop/blob/develop/PACKAGING.md) are up to date
- [x] Verify option parity (make sure that all completions, the manpage, `--help` and getopt provide the same flags)
- [x] Print a warning if positional arguments are supplied
- [x] Think about what should happen when Catch2 is unavailable (related to package management)
- [x] Think about the ways of handling dependencies in CMake

# General testing
- [x] Thoroughly test new history format (test interaction with different formatters etc.)
- [x] Make sure that `--wrapper` works as expected
      `j4-dmenu-desktop --wrapper i3-msg\ exec --skip-i3-exec-check` still works as expected (`-I` should be used instead, but I want to verify that the `--wrapper` approach still works to make sure that I won't break existing configs as much (although I've already broke them with `--skip-i3-exec-check`)).
- [x] #163

# Distro testing
I plan to test the new release on two Linux distros and on FreeBSD. I will test on Void Linux (distro I actively use) and Debian.

## Void Linux
- [x] Test building on glibc variant using Meson
- [x] Test building on glibc variant using CMake
- [x] Test building on musl variant using Meson
- [x] Test building on musl variant using CMake
- [x] Run unit tests on glibc variant
- [x] Run unit tests on musl variant

I will also test the packagebility of j4-dmenu-desktop on Void Linux, because I am familiar with its build system. This will test that building in chroot with fake destdir, cross compiling and more will work as expected.

- Package using Meson
  - [x] `x86_64->x86_64` native
  - [x] `i686->i686` native
  - [x] `x86_64->aarch64` cross
  - [x] `x86_64->armv7l` cross
  - [x] `x86_64-musl->x86_64-musl` native
  - [x] `x86_64-musl->armv6l-musl` cross
  - [x] `x86_64-musl->aarch64-musl` cross
- Package using CMake
  - [x] `x86_64->x86_64` native
  - [x] `i686->i686` native
  - [x] `x86_64->aarch64` cross
  - [x] `x86_64->armv7l` cross
  - [x] `x86_64-musl->x86_64-musl` native
  - [x] `x86_64-musl->armv6l-musl` cross
  - [x] `x86_64-musl->aarch64-musl` cross

## Debian (docker)
- [x] Test building NOTE: debian's Meson is too old (at least the one from the default repositories is), CMake must be used
- [x] Run unit tests

## Debian (VM)
- [x] Test building
- [x] Run unit tests
- [x] Test Notify behavior (run j4-dmenu-desktop in daemon mode, add & remove programs and see whether the daemon is still up to date)
      Fixed in 24523f45bcf49b2aae6b8460a7f280d034421658

## FreeBSD (VM)
- [x] Test building
- [x] Run unit tests
- [x] Test Notify behavior (run j4-dmenu-desktop in daemon mode, add & remove programs and see whether the daemon is still up to date)
      This testing is especially important on FreeBSD because [an alternative implementation of the Notify system is used there](https://github.com/enkore/j4-dmenu-desktop/blob/develop/src/NotifyKqueue.cc)

**Comment by meator** [comment]
I now consider j4-dmenu-desktop to be "production ready". I plan to release j4-dmenu-desktop r3.0 on **the second of April**. I will do some additional minor testing until then.

@enkore I've emailed you detailing my thoughts on the r3.0 release. Feel free to comment on it here.

The finalizing commit should look something like this (do **not** apply before the set release date!):

```diff
diff --git a/CHANGELOG b/CHANGELOG
index 4b4856e..acfbcce 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -80,7 +80,7 @@ r2.18 released 2020-10-02
 	+ added --wrapper option (e.g. for i3 exec)
 	+ removed last usage of C/C++ locales
 
-r3.0 to be released
+r3.0 released 2024-06-02
 	+ BREAKING CHANGE: --i3-ipc should be preferred over --wrapper "i3-msg exec"
 	  j4-dmenu-desktop will fail to start with --wrapper containing i3 unless
 	  --skip-i3-exec-check is passed
diff --git a/version.txt b/version.txt
index fc249e9..9f55b2c 100644
--- a/version.txt
+++ b/version.txt
@@ -1 +1 @@
-2.18
+3.0
```

**Comment by meator** [comment]
Hey @ibotty, would you mind doing some testing of the current HEAD of j4-dmenu-destkop if you have time to spare? Since you've found bugs in j4-dmenu-desktop before, I'd appreciate your help in ironing out any remaining packaging issues for Fedora before the release is made.

I don't plan to make any new breaking changes to j4-dmenu-desktop until the r3.0 release. I will only work on bugfixes and documentation improvements.

**Comment by ibotty** [comment]
I'll start a scratch-build tomorrow and test my use case.  Looking forward to it.

**Comment by meator** [comment]
Because this issue tracks the r3.0 release, I should also make mention of the major breaking change[^1]:

# Breaking changes
i3 support using the old `--wrapper "i3-msg exec"` will no longer work without user intervention. Users should switch to the new i3 IPC integration using `--i3-ipc` instead.

The error can be bypassed using the `--skip-i3-exec-check` flag, although I do not recommend using this flag, because the new i3 integration is faster and it handles escaping properly.

[^1]: I already mention it in [CHANGELOG](https://github.com/enkore/j4-dmenu-desktop/blob/develop/CHANGELOG#L84-L86) and I'll mention it in release notes, but I will put it here too for completeness.

**Comment by ibotty** [comment]
I built scratch built and tested the current HEAD on my Fedora 40 system on sway (Sway Atomic with overlayed package).  It all worked fine.  As soon as you release 3.0 I'll put the new release into the upcoming Fedora version.

**Comment by meator** [comment]
@ibotty Thanks for testing!

**Comment by meator** [comment]
[I have created a prerelease](https://github.com/enkore/j4-dmenu-desktop/releases/tag/r3.0-rc1) of r3.0 to simplify testing.

**Comment by meator** [comment]
I know that I've stated that no new features will be added, but there is a small feature that I want to include in the r3.0 release. I have introduced a `NO_DOWNLOAD` flag to CMake: cba097e1da1563a96e95f5c3d13791a8de39d915

Using

```
cmake -DNO_DOWNLOAD=ON ..
```

is significantly shorter than

```
cmake -DWITH_GIT_CATCH=OFF -DWITH_GIT_SPDLOG=OFF -DWITH_GIT_FMT=OFF ..
```

while achieving the same effect. It is more descriptive and future-proof (if a hypothetical version r3.1 would introduce a new dependency, packagers wanting to use their packages for dependencies would have to add `-DWITH_GIT_<DEPENDENCY>=OFF` to configure args, but with this new approach, `-DNO_DOWNLOAD=ON` will turn off fetching of the new dependency).

I have added this feature because I believe it will simplify package management for distros still relying on the CMake build system for j4-dmenu-desktop (not all distros have Meson >=1.2.0 available).

---

Tomorrow, I will do another round of testing and make the r3.0 release as I've promised. I'm excited to push j4-dmenu-desktop forward! The latest release has been made four years ago; it has bugs and doesn't even build on modern distros. r3.0 should be a breath of fresh air for j4-dmenu-desktop. I've implemented many new features (while trying to adhere to the original vision of j4-dmenu-desktop), fixed bugs and improved documentation.

**Comment by meator** [comment]
I have decided to add another behavior change: 1bbdaed7b342324f78a6466ef5253cf28eb7dedb

Unity builds are now opt-out instead of opt-in for Meson. I was considering this change for quite some time now. I first didn't want to include it in the r3.0 release because of my promise to not change anything and because I viewed setting unity builds by default as [intrusive](https://github.com/enkore/j4-dmenu-desktop/commit/1bbdaed7b342324f78a6466ef5253cf28eb7dedb#diff-f1355fd54329147c830843ab133fed7716180c3242031487b861a2a3235d0078L67-L70). But I now believe the speed benefits outweigh the annoyance with actively developing a project with unity builds.

Unity builds can be disabled with `meson setup --unity off` or using other ways [documented in BUILDING.md](https://github.com/enkore/j4-dmenu-desktop/blob/develop/BUILDING.md#unity-builds-in-meson) This should be pretty straight forward to do for users who prefer normal builds.

j4-dmenu-desktop was already using unity builds before I came in and rewrote the build system & parts of the code. But it was doing it "the hard way". j4-dmenu-desktop was structured like a header only library. This approach has too many drawbacks to list here. My rewrite is a much saner implementation of unity builds. Unity builds can now be toggled, improving the experience for developers who don't want to rebuild everything when changing a single line of code.

The point is that r2.18 already had "unity builds" by default (by "by default" I mean that unity builds were hardcoded in the structure of the codebase), so having unity builds by default is logical for r3.0.

By the way, I have hardcoded unity builds in CMake. I could come up with a more elaborate system for CMake, but I direct most of my attention to Meson now for reasons I've described before. If anyone would find support for "normal" builds in CMake useful, feel free to comment here or submit an issue. Pull request implementing such change are also welcome.
